### PR TITLE
SOC-1788 | Discussion categories editing

### PR DESCRIPTION
**Pull request by kvas-damian**
## Links

* https://wikia-inc.atlassian.net/browse/SOC-1788
* Designs: https://wikia-inc.atlassian.net/browse/SOC-1951

## Description

* Allow admin users to add new categories on discussions
* New UI component - highligh-overlay

## Reviewers

@bkoval @rafalkalinski

### CODE REVIEW

File: front/main/app/components/discussion-categories-edit.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+export default Ember.Component.extend(ResponsiveMixin,
+	{
+		classNames: ['highlight-overlay-content', 'discussion-categories-edit'],
+		maxCategoriesCount: 10,
+		isLoading: false,
+		showSuccess: false,
+		wikiId: Ember.get(Mercury, 'wiki.id').toString(),
+
+		addDisabled: Ember.computed('localCategories.length', function () {
+			return this.get('localCategories.length') >= this.get('maxCategoriesCount');
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories-edit.js]: if the localCategories count is equal maxCategoriesCount, do we really need to set addDisabled to ```true```?

File: front/main/app/components/discussion-categories-edit.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		}),
+
+		localCategories: Ember.computed('categories.@each', function () {
+			const categories = this.get('categories'),
+				localCategories = new Ember.A();
+
+			categories.forEach((category) => {
+				localCategories.pushObject(Ember.Object.create(category));
+			});
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories-edit.js]: ```localCategories = categories.map((category) => Ember.Object.create(category));```

File: front/main/app/components/discussion-categories-edit.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+		actions: {
+			/**
+			 * Add new category
+			 *
+			 * @returns {void}
+			 */
+			addCategory() {
+				if (!this.get('addDisabled')) {
+					this.get('localCategories').pushObject(DiscussionCategory.create({}));
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories-edit.js]: Does it add an empty category to the list in UI? It is not obvious without a comment.

File: front/main/app/components/discussion-categories-edit.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+			/**
+			 * Delete a category
+			 *
+			 * @param {DiscussionCategory} category category to delete
+			 *
+			 * @returns {void}
+			 */
+			deleteCategory(category) {
+				this.get('localCategories').removeObject(category);
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories-edit.js]: tracking?

File: front/main/app/components/discussion-categories.js [hunk 5]
```diff
@@ -189,8 +195,23 @@ export default Ember.Component.extend(
...
+			},
+
+			/**
+			 * Enables/disables categories edit mode
+			 *
+			 * @param {boolean} isEnabled edit mode state
+			 *
+			 * @returns {void}
+			 */
+			setEditMode(isEnabled) {
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories.js]: Bart would probably say: ```shouldEnabled``` :)

File: front/main/app/components/discussion-categories.js [hunk 6]
```diff
@@ -8,6 +8,12 @@ export default Ember.Component.extend(
 		disabled: false,
```
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories.js]: This module takes over 5 pages on my screen.
Maybe it's the time to think about splitting it into parts?

File: front/main/app/components/wikia-ui-components/highlight-overlay.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+
+/**
+ * Fancy-check-box component usage:
```
Comment (rafalkalinski) [inline front/main/app/components/wikia-ui-components/highlight-overlay.js]: Please, fix that comment.

File: front/main/app/models/discussion/categories.js [hunk 8]
```diff
@@ -43,16 +43,44 @@ const DiscussionModel = Ember.Object.extend({
...
+
+			this.set('categories', categories.sortBy('displayOrder'));
+		});
+	},
+
+	updateCategories(categories) {
+		const promisesList = [];
+
+		categories.rejectBy('id').forEach((category) => {
+			promisesList.push(this.addCategory(category.get('name')));
```
Comment (rafalkalinski) [inline front/main/app/models/discussion/categories.js]: also here - using .map method would save some performance

File: front/main/app/models/user-permissions/domain/discussions.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	/**
+	 * @param {Object} permissionsData
+	 *
+	 * @returns {Ember.Object}
+	 */
+	create(permissionsData) {
+		const permissions = {};
+
+		Object.keys(permissionsMap).forEach((permission) => {
+			if (permissionsData.hasOwnProperty(permissionsMap[permission]) || permissionsData.indexOf('*') !== -1) {
```
Comment (rafalkalinski) [inline front/main/app/models/user-permissions/domain/discussions.js]: why a ```*```?
beside that - I love the solution with the two objects here :100: :+1:

File: front/main/app/models/user.js [hunk 10]
```diff
@@ -51,7 +69,7 @@ UserModel.reopenClass({
 			data: {
 				controller: 'UserApi',
 				method: 'getDetails',
-				ids: userId,
+				ids: 1234,
```
Comment (rafalkalinski) [inline front/main/app/models/user.js]: some stag data left?

File: front/main/app/models/user.js [hunk 11]
```diff
@@ -82,7 +130,71 @@ UserModel.reopenClass({
...
+	},
+
+	/**
+	 * @param {Ember.Object} model
+	 * @param {QueryUserInfoResponse} query
+	 * @returns {void}
+	 */
+	setUserRights(model, {query}) {
+		const rightsArray = query.userinfo.rights,
+			rights = {};
```
Comment (rafalkalinski) [inline front/main/app/models/user.js]: ABC order

File: front/main/app/models/user.js [hunk 12]
```diff
@@ -82,7 +130,71 @@ UserModel.reopenClass({
...
+
+		if (Ember.isArray(rightsArray)) {
+			// TODO - we could use contains instead of making an object out of an array
```
Comment (rafalkalinski) [inline front/main/app/models/user.js]: and what about this comment?

File: front/main/app/styles/component/_discussion-categories-edit.scss [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+	}
+
+	.discussion-categories-edit-action-button {
+		position: absolute;
+		right: 0;
+	}
+
+	.icon-submit {
+		width: 35px;
+		height: 35px;
```
Comment (rafalkalinski) [inline front/main/app/styles/component/_discussion-categories-edit.scss]: Bart's comment.

File: front/main/app/styles/component/_discussion-categories.scss [hunk 14]
```diff
@@ -3,6 +3,10 @@
 	max-height: 800px;
 	transition: max-height .6s;
 	overflow: hidden;
+
+	.discussion-categories-edit & {
```
Comment (rafalkalinski) [inline front/main/app/styles/component/_discussion-categories.scss]: what's up with the ```&```?

File: front/main/app/templates/components/discussion-categories.hbs [hunk 15]
```diff
@@ -1,15 +1,30 @@
...
+		localCategories=localCategories
+		selectAllCategory=(action 'selectAllCategory')
+		onCategoryClick=(action 'onCategoryClick')
+		reset=(action 'reset')
+		setEditMode=(action 'setEditMode')
+		toggle=(action 'toggle')
+		toggleButtonLabel=toggleButtonLabel
+		toggleButtonVisible=toggleButtonVisible
+		toggleMore=(action 'toggleMore')
+		visibleCategoriesCount=visibleCategoriesCount
```
Comment (rafalkalinski) [inline front/main/app/templates/components/discussion-categories.hbs]: order

File: front/main/app/templates/components/discussion-entities-list-layout.hbs [hunk 16]
```diff
@@ -8,6 +8,7 @@
 		hasNewPostButton=true
 		hasSiteName=true
 		smartBannerVisible=smartBannerVisible
+		updateCategories=(action this.attrs.updateCategories)
```
Comment (rafalkalinski) [inline front/main/app/templates/components/discussion-entities-list-layout.hbs]: I really hate comments about: order

File: front/main/app/templates/components/discussion-header.hbs [hunk 17]
```diff
@@ -27,6 +27,7 @@
 					showSortSection=true
 					canModerate=canModerate
 					categories=categories
+					updateCategories=(action this.attrs.updateCategories)
```
Comment (rafalkalinski) [inline front/main/app/templates/components/discussion-header.hbs]: order

File: front/main/app/components/discussion-categories-edit.js (hunk 4 above)
Comment (bkoval) [inline front/main/app/components/discussion-categories-edit.js]: Deleting is not a part of this story, this is just a bit of overhead functionality added here.

File: front/main/app/components/discussion-categories-edit.js (hunk 3 above)
Comment (bkoval) [inline front/main/app/components/discussion-categories-edit.js]: It's an edit mode, so it's just adding a new input for category below all existing inputs.

File: front/main/app/components/discussion-categories.js (hunk 6 above)
Comment (bkoval) [inline front/main/app/components/discussion-categories.js]: This is 217 lines, we have much bigger components in our codebase. How about we proceed with it and when we feel there's some parts we can extract we can do it later on, while adding functionality to this module?

File: front/main/app/components/discussion-categories.js (hunk 5 above)
Comment (bkoval) [inline front/main/app/components/discussion-categories.js]: I'd rather say shouldEnable :)

File: front/main/app/components/discussion-categories-edit.js (hunk 1 above)
Comment (kvas-damian) [inline front/main/app/components/discussion-categories-edit.js]: I don't understand why we shouldn't? I allow up to 10 categories to be created by user, when user reaches limit of 10 the button gets disabled.

File: front/main/app/components/discussion-categories-edit.js (hunk 4 above)
Comment (kvas-damian) [inline front/main/app/components/discussion-categories-edit.js]: It's not fully an overhead, but this action allow user to remove category that he just add. Not sure if we should implement tracking here.

File: front/main/app/models/user-permissions/domain/discussions.js (hunk 9 above)
Comment (bkoval) [inline front/main/app/models/user-permissions/domain/discussions.js]: An asterisk is a "god mode" returned from the service for the staff users. It gives you all permissions at once. We'd much rather have just all permissions returned explicitly, but that's what we have coming from backend ;(

File: front/main/app/templates/components/discussion-categories.hbs (hunk 15 above)
Comment (bkoval) [inline front/main/app/templates/components/discussion-categories.hbs]: addressed

File: front/main/app/components/discussion-categories-edit.js (hunk 1 above)
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories-edit.js]: that's ok, At the beginning I thought it was for something else.

File: front/main/app/templates/components/discussion-entities-list-layout.hbs (hunk 16 above)
Comment (kvas-damian) [inline front/main/app/templates/components/discussion-entities-list-layout.hbs]: I'm not sure we use the same alphabet...

File: front/main/app/components/discussion-categories.js (hunk 6 above)
Comment (rafalkalinski) [inline front/main/app/components/discussion-categories.js]: I'm totally okay with this approach as soon as we're honest with each other:
1. We've done worse
2. We're going to force someone else to refactor this module next time we're doing CR here

Comment (kvas-damian) [inline front/main/app/components/discussion-categories.js]: I'd go with the latter ;)

File: front/main/app/templates/components/discussion-entities-list-layout.hbs (hunk 16 above)
Comment (rafalkalinski) [inline front/main/app/templates/components/discussion-entities-list-layout.hbs]: nevermind :-1:

**Comment by rafalkalinski** [comment]
LGTM

**Comment by bkoval** [comment]
lgtm 🔢

File: front/main/package.json [hunk 18]
```diff
@@ -48,6 +48,7 @@
     "ember-cli-uglify": "1.2.0",
     "ember-derequire": "0.1.2",
     "ember-disable-proxy-controllers": "1.0.1",
+    "ember-drag-drop": "0.3.6",
```
Comment (rafalkalinski) [inline front/main/package.json]: please, make sure if shrink-wrap is ready with this package

File: front/main/app/models/discussion/categories.js [hunk 19]
```diff
@@ -124,21 +143,48 @@ const DiscussionCategoriesModel = Ember.Object.extend({
 		});
 	},
 
+	getDeletedCategories(categories) {
+		const leftCategoryIds = categories.mapBy('id');
+
+		return this.get('categories').filter((category) => {
+			return leftCategoryIds.indexOf(category.get('id')) === -1;
```
Comment (rafalkalinski) [inline front/main/app/models/discussion/categories.js]: ```return leftCategoryIds.includes(category.get('id'));```

**Comment by rafalkalinski** [comment]
Make sure the test are passing

**Comment by rafalkalinski** [comment]
Please remember about a localization ticket

**Pull request closed without merging**